File: Sources/UsageWidget/Domain/LimitWindow.swift
import Foundation

/// A single rate-limit window for a provider, normalized into a common shape.
///
/// Both providers report a short rolling window (~5 hours) and a long window (~weekly).
/// Some providers (Claude) additionally break the weekly window down per model family.
struct LimitWindow: Identifiable, Hashable, Codable, Sendable {
    enum Kind: String, Codable, Sendable, CaseIterable {
        case fiveHour
        case weekly
        case weeklyOpus
        case weeklySonnet

        /// Short, human-facing title for the bar.
        var title: String {
            switch self {
            case .fiveHour: return "5-hour"
            case .weekly: return "Weekly"
            case .weeklyOpus: return "Weekly · Opus"
            case .weeklySonnet: return "Weekly · Sonnet"
            }
        }

        /// The two windows that make up the hero view. Others fold into "Details".
        var isHero: Bool { self == .fiveHour || self == .weekly }

        /// Sort order so `5-hour` is always shown above `Weekly`.
        var sortIndex: Int {
            switch self {
            case .fiveHour: return 0
            case .weekly: return 1
            case .weeklyOpus: return 2
            case .weeklySonnet: return 3
            }
        }
    }

    let kind: Kind
    /// Normalized fraction used, 0...1. Providers report 0-100 percentages; map by /100.
    let utilization: Double
    /// When this window rolls over. Optional because a provider may omit it.
    let resetsAt: Date?

    var id: Kind { kind }

    init(kind: Kind, utilization: Double, resetsAt: Date?) {
        self.kind = kind
        self.utilization = utilization
        self.resetsAt = resetsAt
    }

    /// Utilization clamped to a safe 0...1 for display math.
    var clampedUtilization: Double { min(max(utilization, 0), 1) }

    /// Integer percentage used, e.g. 47.
    var percent: Int { Int((clampedUtilization * 100).rounded()) }
}

File: windows/src/Core/Domain/LimitWindow.cs
using LLMUsageWidget.Core.Support;

namespace LLMUsageWidget.Core.Domain;

public enum LimitWindowKind { FiveHour, Weekly, WeeklyOpus, WeeklySonnet, Monthly }

public static class LimitWindowKindExtensions
{
    /// <summary>Default human-facing title (a window may override via <c>Label</c>).</summary>
    public static string Title(this LimitWindowKind k) => k switch
    {
        LimitWindowKind.FiveHour => "5-hour",
        LimitWindowKind.Weekly => "Weekly",
        LimitWindowKind.WeeklyOpus => "Weekly · Opus",
        LimitWindowKind.WeeklySonnet => "Weekly · Sonnet",
        LimitWindowKind.Monthly => "Monthly",
        _ => k.ToString()
    };

    /// <summary>Windows that make up the hero view. Others fold into "Details".</summary>
    public static bool IsHero(this LimitWindowKind k) =>
        k is LimitWindowKind.FiveHour or LimitWindowKind.Weekly or LimitWindowKind.Monthly;

    public static int SortIndex(this LimitWindowKind k) => k switch
    {
        LimitWindowKind.FiveHour => 0,
        LimitWindowKind.Monthly => 1,
        LimitWindowKind.Weekly => 2,
        LimitWindowKind.WeeklyOpus => 3,
        LimitWindowKind.WeeklySonnet => 4,
        _ => 99
    };
}

/// <summary>A single usage window for a provider, normalized into a common shape. Rolling windows
/// (Claude/Codex 5-hour + weekly) and quota-style windows (Copilot monthly) both use this.</summary>
public sealed record LimitWindow(
    LimitWindowKind Kind,
    double Utilization,
    DateTimeOffset? ResetsAt,
    string? Label = null,
    double? Used = null,
    double? Limit = null,
    bool Unlimited = false)
{
    public string DisplayTitle => Label ?? Kind.Title();

    /// <summary>Utilization clamped to a safe 0..1 for display math.</summary>
    public double ClampedUtilization => Math.Min(Math.Max(Utilization, 0), 1);

    /// <summary>Integer percentage used, e.g. 47.</summary>
    public int Percent => (int)Math.Round(ClampedUtilization * 100, MidpointRounding.AwayFromZero);

    /// <summary>"173 / 1.5K" when raw counts are present.</summary>
    public string? CountText
    {
        get
        {
            if (Used is not double u || Limit is not double l || l <= 0) return null;
            int used = (int)Math.Round(u, MidpointRounding.AwayFromZero);
            int limit = (int)Math.Round(l, MidpointRounding.AwayFromZero);
            return $"{NumberFormat.Compact(used)} / {NumberFormat.Compact(limit)}";
        }
    }
}
